### ISSUE TITLE
Message Passing : Extremely Slow backpropagation speed

### ISSUE THREAD

**Issue by PlayeerOne** [pyg-team/pytorch_geometric#10011]
Hi all, 

I have implemented my own version of HGTConv which is heavily based on the torch_geometric's implementation, I am working on hetergeneous graphs, and the number of node types and edge types present on a graph differ from one graph to another. In fact I made my own implementation because I don't want to take into account the whole edge triplets (src_type, rel_type, dst_type) but rather just the relation type. 
My implementation works fine but the backpropagation takes a lot of time (about 2 minutes for a batch of size 24), I have tried limiting moving matrix slices around but it dont seem to work. 

### Versions

I am using torch_geometric 2.7.0

**Comment by puririshi98** [comment]
few questions before i can help:
what dataset are you using?

what is the size of the model?

what computer are you using?

it sounds like your code may not be using your gpu properly

**Comment by PlayeerOne** [comment]
Hi @puririshi98,

Thanks for your response!

I'm running on a 24GB GPU and using a custom dataset for a graph-level prediction task. The key details:

- Node types and edge types vary per graph
- Total edge types (relationship types): ~140
- Total node types: ~70

Due to the large number of possible `(src_type, rel_type, dst_type)` combinations, I implemented a custom `HGTConv `that only considers `rel_type `for computing the q, k, v matrices instead of taking into account the whole (src_type, rel_type, dst_type)` triplet.

My model consists of:

- Two custom `HGTConv `layers (hidden size: 512, 16 heads)
- Projection + norm layers in between
- Classification head

I profiled the backward pass and found that computing the k and v matrices is responsible for a significant portion of the compute time during backprop. However, inference is fast.

Given the slicing and movement of matrices in the original `HGTConv `implementation (on which I based my implementation on), I suspect that the high number of relationship and node types is causing inefficiencies—particularly because PyG doesn’t fuse gradients, and this makes the computation graph very large, leading to slower backpropagation.

Would love any insights on optimizing this!

**Comment by puririshi98** [comment]
great analysis, since you are aware of the inner workings, i would reocmmend reproducing the issue in a seperate python file that relies solely on pytorch. with that it would be much easier for me to help unblock u

**Comment by PlayeerOne** [comment]
I will do that as soon as possible, it will take a while for me to provide the code which reproduces the phenomena because it is on a private project.

**Comment by puririshi98** [comment]
i will close this for now, please open a new issue citing the reproduction steps needed to witness the slowdown using the NVIDIA container and please list your hardware specs as well (CPU and GPU)
